plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE ConstraintKinds  #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE TypeFamilies     #-}
+
+{-|
+Call site inlining machinery. For now there's only one part: inlining of fully applied functions.
+See ADR TODO for motivation. We inline fully applied functions if the cost and size are acceptable.
+See note [Inlining of fully applied functions].
+-}
+
+module PlutusIR.Transform.Inline.CallSiteInline where
+
+{- Note [Inlining of fully applied functions]
+
+We inline the call site when a function is fully applied AND the cost and size are acceptable.
+
+Consider `let v :: ty = rhs in body`, in which `body` calls `v`.
+
+We consider cases when `v` is a function/lambda abstraction(s). I.e.:
+
+let v :: type = \x1.\x2...\xn.VBody in body
+
+In the `body`, where `v` is *called*,
+if it was given `n` arguments, then it is fully applied in the `body`.
+We inline the *call site* of the fully applied `v` in this case, i.e.,
+we replace `v x_1 ... x_n` in the `body` with `rhs`. E.g.
+
+let f = \x.\y -> x
+in
+  let z = f q
+  in f a b
+
+becomes
+
+let f = \x.\y -> x
+in
+  let z = f q
+  in a
+
+because f is fully applied in the `body`. Our dead code elimination pass should then turn it to`a`,
+reducing the code size. What about
+
+let f = \x.\y -> x
+    z = f q
+in f a b + z c
+
+Here we have a `f` that is fully applied (`f a b`), and so we inline it:
+
+let f = \x.\y -> x
+in
+  let z = f q
+  in a + z c
+
+We cannot eliminate the let-binding of `f` because it is not dead (it's called in `z`).
+We don't inline `z` because it's not a lambda abstraction, it's an application.
+
+To do this we need to count the number of type/term lambda abstractions,
+so we know how many term/type arguments they have.
+
+We pattern match the _rhs_ with `LamAbs` or `TyAbs` (lambda abstraction for terms or types),
+and count the number of lambdas.
+Then, in the _body_, we check for term or type application (`Apply` or `TyInst`) of _v_.
+
+If _v_ is fully applied in the body, i.e., if
+
+1. the number of type lambdas equals the number of type arguments applied, and
+2. the number of term lambdas equals the number of term arguments applied, and
+
+if other call site inlining conditions are satisfied,
+(we currently inline too little, this will be improved later, see below)
+
+we inline _v_, i.e., replace its occurrence with _rhs_ in the _body_.
+For the rest of the discussion here we focus on the conditions 1 and 2 only.
+
+Below are some examples that involve a _body_ that is not fully reducible/applied but
```

Comment:
What does this mean? I still want a specific discussion of the various concepts of reducible/saturated, I think it's _critical_.